File: Engine/Sources/CardCopilotEngine/Engine/RecommendationEngine.swift
import Foundation

public struct Recommendation: Equatable, Sendable {
    public let winner: CandidateScore
    public let runnerUp: CandidateScore?
    public let switchedFromDefault: Bool
    public let advantageOverDefaultCad: Double?
    public let defaultNotAccepted: Bool
    /// A card that beat the default but not by enough to be worth digging out the wallet.
    public let suppressedBetterCard: CandidateScore?
    /// True when the winner depends on the owner's declared point valuation — valuing points
    /// lower (or higher) would pick a different card.
    public let valuationSensitive: Bool
    /// Which way the declared valuation would have to move to change the advice.
    public let valuationDirection: ValuationDirection?
    /// The card that wins on the other side of the breakeven.
    public let alternateWinnerCardId: String?
    /// The cents-per-point at which the recommendation flips to `alternateWinnerCardId`.
    public let breakevenCentsPerPoint: Double?
    /// The declared cents-per-point the winning score assumed, when valuation-sensitive.
    public let declaredCentsPerPoint: Double?
    public let allCandidates: [CandidateScore]
}

/// Which direction the point valuation would have to move for the advice to change.
public enum ValuationDirection: Sendable, Equatable { case below, above }

public struct RecommendationEngine {
    let catalogue: Catalogue
    let ownerState: OwnerState

    /// Catalogue valuation defaults are merged in here, beneath anything the owner has declared.
    /// This is the single funnel every scoring path reaches — including owner states restored
    /// from a device, which never touch SeedLoader. Without it contracts/programs.json would be
    /// data nothing reads.
    public init(catalogue: Catalogue, ownerState: OwnerState) {
        self.catalogue = catalogue
        self.ownerState = ownerState.applyingCatalogueValuationDefaults()
    }

    private struct Verdict {
        let winner: CandidateScore
        let runnerUp: CandidateScore?
        let switched: Bool
        let advantage: Double?
        let defaultNotAccepted: Bool
        let suppressed: CandidateScore?
        let ranked: [CandidateScore]
    }

    public func recommend(_ purchase: PurchaseContext, asOf: String) -> Recommendation {
        let candidateCards = ownerState.ownedCardIds.isEmpty
            ? catalogue.cards
            : catalogue.cards.filter { ownerState.ownedCardIds.contains($0.cardId) }
        let scores = candidateCards
            .map { Scorer.score(card: $0, purchase: purchase, ownerState: ownerState, asOf: asOf) }
            .filter { !$0.excluded }
        precondition(!scores.isEmpty, "no scorable card — catalogue misconfigured")

        let declared = rank(scores, purchase: purchase, value: { $0.netValueCad })
        let floor = rank(scores, purchase: purchase, value: { $0.floorNetValueCad })
        let aspirational = rank(scores, purchase: purchase, value: { $0.aspirationalNetValueCad })

        var sensitive = false
        var direction: ValuationDirection?
        var alternateId: String?
        var breakeven: Double?
        var declaredCents: Double?

        // Downside: the winner is a points card that only wins because points are declared
        // above their guaranteed floor.
        if declared.winner.cardId != floor.winner.cardId,
           abs(declared.winner.floorNetValueCad - declared.winner.netValueCad) > 0.0001,
           declared.winner.rewardUnits > 0 {
            sensitive = true
            direction = .below
            alternateId = floor.winner.cardId
            breakeven = breakevenCents(pointsCard: declared.winner,
                                       incumbent: floor.winner,
                                       ranked: declared.ranked, purchase: purchase)
            declaredCents = centsPerUnit(declared.winner)
        }
        // Upside: a points card would overtake the winner if points were worth more. Only
        // disclosed when the flip happens within the published benchmark — past that it is
        // noise, not information.
        else if aspirational.winner.cardId != declared.winner.cardId,
                aspirational.winner.rewardUnits > 0,
                abs(aspirational.winner.aspirationalNetValueCad
                    - aspirational.winner.netValueCad) > 0.0001,
                let challenger = declared.ranked.first(where: { $0.cardId == aspirational.winner.cardId }) {
            let flip = breakevenCents(pointsCard: challenger,
                                      incumbent: declared.winner,
                                      ranked: declared.ranked, purchase: purchase)
            let benchmarkCents = (challenger.aspirationalNetValueCad + challenger.fxCostCad)
                * 100 / challenger.rewardUnits
            if flip <= benchmarkCents + 0.0001 {
                sensitive = true
                direction = .above
                alternateId = challenger.cardId
                breakeven = flip
                // The disclosed value is the challenger's currency — that is the number the
                // owner would be revising, not the cash-back winner's notional "unit" value.
                declaredCents = centsPerUnit(challenger)
            }
        }

        return Recommendation(winner: declared.winner,
                              runnerUp: declared.runnerUp,
                              switchedFromDefault: declared.switched,
                              advantageOverDefaultCad: declared.advantage,
                              defaultNotAccepted: declared.defaultNotAccepted,
                              suppressedBetterCard: declared.suppressed,
                              valuationSensitive: sensitive,
                              valuationDirection: direction,
                              alternateWinnerCardId: alternateId,
                              breakevenCentsPerPoint: breakeven,
                              declaredCentsPerPoint: declaredCents,
                              allCandidates: declared.ranked)
    }

    private func centsPerUnit(_ score: CandidateScore) -> Double {
        score.rewardUnits > 0 ? (score.grossRewardCad / score.rewardUnits) * 100 : 0
    }

    /// The cents-per-point at which `pointsCard` and `incumbent` swap places, accounting for
    /// the switch threshold that applies against the default card. Cross-validated against
    /// bisection over the full engine in BreakevenCrossValidationTests.
    private func breakevenCents(pointsCard: CandidateScore, incumbent: CandidateScore,
                                ranked: [CandidateScore], purchase: PurchaseContext) -> Double {
        let t = ownerState.switchThreshold
        let ppFloorCad = t.minAdvantagePercentagePoints * purchase.amountCad / 100
        let requiredAdvantage = t.semantics == "either"
            ? min(t.minAdvantageCad, ppFloorCad)
            : max(t.minAdvantageCad, ppFloorCad)
        let defaultId = ownerState.defaultCardId

        // The points card must clear the incumbent, plus the switch threshold over the default.
        var needed = incumbent.netValueCad
            + (incumbent.cardId == defaultId ? requiredAdvantage : 0)
        if incumbent.cardId != defaultId, pointsCard.cardId != defaultId,
           let defaultScore = ranked.first(where: { $0.cardId == defaultId }) {
            needed = max(needed, defaultScore.netValueCad + requiredAdvantage)
        }
        return (needed + pointsCard.fxCostCad) * 100 / pointsCard.rewardUnits
    }

    private func rank(_ scores: [CandidateScore], purchase: PurchaseContext,
                      value: (CandidateScore) -> Double) -> Verdict {
        let defaultId = ownerState.defaultCardId
        let ranked = scores.sorted { a, b in
            if value(a) != value(b) { return value(a) > value(b) }
            if a.cardId == defaultId { return true }
            if b.cardId == defaultId { return false }
            return a.cardId < b.cardId
        }

        let best = ranked[0]
        let runnerUp = ranked.count > 1 ? ranked[1] : nil

        guard let defaultScore = ranked.first(where: { $0.cardId == defaultId }) else {
            return Verdict(winner: best, runnerUp: runnerUp, switched: true, advantage: nil,
                           defaultNotAccepted: true, suppressed: nil, ranked: ranked)
        }

        let advantage = value(best) - value(defaultScore)
        let advantagePP = purchase.amountCad > 0 ? advantage / purchase.amountCad * 100 : 0
        let t = ownerState.switchThreshold
        let cadOk = advantage >= t.minAdvantageCad
        let ppOk = advantagePP >= t.minAdvantagePercentagePoints
        let clearsThreshold = t.semantics == "either" ? (cadOk || ppOk) : (cadOk && ppOk)

        if best.cardId != defaultId && clearsThreshold {
            return Verdict(winner: best, runnerUp: runnerUp, switched: true, advantage: advantage,
                           defaultNotAccepted: false, suppressed: nil, ranked: ranked)
        }

        let suppressed = (best.cardId != defaultId && advantage > 0) ? best : nil
        return Verdict(winner: defaultScore,
                       runnerUp: ranked.first { $0.cardId != defaultId },
                       switched: false, advantage: 0,
                       defaultNotAccepted: false, suppressed: suppressed, ranked: ranked)
    }
}

File: android/core/engine/src/main/kotlin/com/cardcopilot/engine/engine/RecommendationEngine.kt
package com.cardcopilot.engine.engine

import com.cardcopilot.engine.models.CandidateScore
import com.cardcopilot.engine.models.Catalogue
import com.cardcopilot.engine.models.OwnerState
import com.cardcopilot.engine.models.PurchaseContext
import com.cardcopilot.engine.models.Recommendation
import com.cardcopilot.engine.models.ValuationDirection
import kotlin.math.abs

class RecommendationEngine(
    val catalogue: Catalogue,
    val ownerState: OwnerState
) {
    private data class Verdict(
        val winner: CandidateScore,
        val runnerUp: CandidateScore?,
        val switched: Boolean,
        val advantage: Double?,
        val defaultNotAccepted: Boolean,
        val suppressed: CandidateScore?,
        val ranked: List<CandidateScore>
    )

    fun recommend(purchase: PurchaseContext, asOf: String): Recommendation {
        val candidateCards = if (ownerState.ownedCardIds.isEmpty()) {
            catalogue.cards
        } else {
            val ownedSet = ownerState.ownedCardIds.toSet()
            catalogue.cards.filter { ownedSet.contains(it.cardId) }
        }
        val scores = candidateCards
            .map { Scorer.score(it, purchase, ownerState, asOf) }
            .filter { !it.excluded }
        check(scores.isNotEmpty()) { "no scorable card — catalogue misconfigured" }

        val declared = rank(scores, purchase) { it.netValueCad }
        val floor = rank(scores, purchase) { it.floorNetValueCad }
        val aspirational = rank(scores, purchase) { it.aspirationalNetValueCad }

        var sensitive = false
        var direction: ValuationDirection? = null
        var alternateId: String? = null
        var breakeven: Double? = null
        var declaredCents: Double? = null

        if (declared.winner.cardId != floor.winner.cardId &&
            abs(declared.winner.floorNetValueCad - declared.winner.netValueCad) > 0.0001 &&
            declared.winner.rewardUnits > 0
        ) {
            sensitive = true
            direction = ValuationDirection.BELOW
            alternateId = floor.winner.cardId
            breakeven = breakevenCents(
                pointsCard = declared.winner,
                incumbent = floor.winner,
                ranked = declared.ranked,
                purchase = purchase
            )
            declaredCents = centsPerUnit(declared.winner)
        } else if (aspirational.winner.cardId != declared.winner.cardId &&
            aspirational.winner.rewardUnits > 0 &&
            abs(aspirational.winner.aspirationalNetValueCad - aspirational.winner.netValueCad) > 0.0001
        ) {
            val challenger = declared.ranked.firstOrNull { it.cardId == aspirational.winner.cardId }
            if (challenger != null) {
                val flip = breakevenCents(
                    pointsCard = challenger,
                    incumbent = declared.winner,
                    ranked = declared.ranked,
                    purchase = purchase
                )
                val benchmarkCents = (challenger.aspirationalNetValueCad + challenger.fxCostCad) * 100.0 / challenger.rewardUnits
                if (flip <= benchmarkCents + 0.0001) {
                    sensitive = true
                    direction = ValuationDirection.ABOVE
                    alternateId = challenger.cardId
                    breakeven = flip
                    declaredCents = centsPerUnit(challenger)
                }
            }
        }

        return Recommendation(
            winner = declared.winner,
            runnerUp = declared.runnerUp,
            switchedFromDefault = declared.switched,
            advantageOverDefaultCad = declared.advantage,
            defaultNotAccepted = declared.defaultNotAccepted,
            suppressedBetterCard = declared.suppressed,
            valuationSensitive = sensitive,
            valuationDirection = direction,
            alternateWinnerCardId = alternateId,
            breakevenCentsPerPoint = breakeven,
            declaredCentsPerPoint = declaredCents,
            allCandidates = declared.ranked
        )
    }

    private fun centsPerUnit(score: CandidateScore): Double {
        return if (score.rewardUnits > 0) (score.grossRewardCad / score.rewardUnits) * 100.0 else 0.0
    }

    private fun breakevenCents(
        pointsCard: CandidateScore,
        incumbent: CandidateScore,
        ranked: List<CandidateScore>,
        purchase: PurchaseContext
    ): Double {
        val t = ownerState.switchThreshold
        val ppFloorCad = t.minAdvantagePercentagePoints * purchase.amountCad / 100.0
        val requiredAdvantage = if (t.semantics == "either") {
            minOf(t.minAdvantageCad, ppFloorCad)
        } else {
            maxOf(t.minAdvantageCad, ppFloorCad)
        }
        val defaultId = ownerState.defaultCardId

        var needed = incumbent.netValueCad + (if (incumbent.cardId == defaultId) requiredAdvantage else 0.0)
        if (incumbent.cardId != defaultId && pointsCard.cardId != defaultId) {
            val defaultScore = ranked.firstOrNull { it.cardId == defaultId }
            if (defaultScore != null) {
                needed = maxOf(needed, defaultScore.netValueCad + requiredAdvantage)
            }
        }
        return (needed + pointsCard.fxCostCad) * 100.0 / pointsCard.rewardUnits
    }

    private fun rank(
        scores: List<CandidateScore>,
        purchase: PurchaseContext,
        value: (CandidateScore) -> Double
    ): Verdict {
        val defaultId = ownerState.defaultCardId
        val ranked = scores.sortedWith { a, b ->
            val va = value(a)
            val vb = value(b)
            if (va != vb) return@sortedWith vb.compareTo(va)
            if (a.cardId == defaultId) return@sortedWith -1
            if (b.cardId == defaultId) return@sortedWith 1
            a.cardId.compareTo(b.cardId)
        }

        val best = ranked[0]
        val runnerUp = if (ranked.size > 1) ranked[1] else null

        val defaultScore = ranked.firstOrNull { it.cardId == defaultId }
            ?: return Verdict(
                winner = best,
                runnerUp = runnerUp,
                switched = true,
                advantage = null,
                defaultNotAccepted = true,
                suppressed = null,
                ranked = ranked
            )

        val advantage = value(best) - value(defaultScore)
        val advantagePP = if (purchase.amountCad > 0) advantage / purchase.amountCad * 100.0 else 0.0
        val t = ownerState.switchThreshold
        val cadOk = advantage >= t.minAdvantageCad
        val ppOk = advantagePP >= t.minAdvantagePercentagePoints
        val clearsThreshold = if (t.semantics == "either") (cadOk || ppOk) else (cadOk && ppOk)

        if (best.cardId != defaultId && clearsThreshold) {
            return Verdict(
                winner = best,
                runnerUp = runnerUp,
                switched = true,
                advantage = advantage,
                defaultNotAccepted = false,
                suppressed = null,
                ranked = ranked
            )
        }

        val suppressed = if (best.cardId != defaultId && advantage > 0) best else null
        return Verdict(
            winner = defaultScore,
            runnerUp = ranked.firstOrNull { it.cardId != defaultId },
            switched = false,
            advantage = 0.0,
            defaultNotAccepted = false,
            suppressed = suppressed,
            ranked = ranked
        )
    }
}
